include/slate/BaseMatrix.hh

```diff
@@ -1283,34 +1284,43 @@ void swap(BaseMatrix<scalar_t>& A, BaseMatrix<scalar_t>& B)
 ///
 template <typename scalar_t>
 void BaseMatrix<scalar_t>::gridinfo(
-    GridOrder* order, int* nprow, int* npcol, int* myrow, int* mycol ) const
-{
-    if (nprow_ > 0) {
-        *order = order_;
-        *nprow = nprow_;
-        *npcol = npcol_;
-        if (order_ == GridOrder::Col) {
-            *myrow = mpi_rank_ % nprow_;
-            *mycol = mpi_rank_ / nprow_;
-        }
-        else {
-            *myrow = mpi_rank_ / npcol_;
-            *mycol = mpi_rank_ % npcol_;
-        }
-    }
-    else {
+    GridOrder* order, int* nprow, int* npcol, int* myrow, int* mycol )
+{
+    // If grid order is unknown, attempt to detect it
+    if (order_ == GridOrder::Unknown) {
         int mpi_size;
-        slate_mpi_call(
-            MPI_Comm_size(mpiComm(), &mpi_size));
-        if (mpi_size == 1) {
-            *order = GridOrder::Col;
-            *nprow = *npcol = 1;
-            *myrow = *mycol = 0;
-        }
-        else {
-            *order = GridOrder::Unknown;
-            *nprow = *npcol = *myrow = *mycol = -1;
+        slate_mpi_call(MPI_Comm_size(mpiComm(), &mpi_size));
+        // When matrix is empty, storage_ may be null
+        if (mt_ != 0 && nt_ != 0) {
+            int64_t nprow_i64, npcol_i64;
+            func::is_2d_cyclic_grid(mt(), nt(), tileRankFunc(), order_, nprow_i64, npcol_i64);
+            nprow_ = nprow_i64;
+            npcol_ = npcol_i64;
```

Comment:
Is there a reason `is_2d_cyclic_grid` takes p and q as `int64_t` instead of `int`?
The motivation for using `int` for MPI grids is that `MPI_Comm_size` takes `int`. Similarly for `cudaGetDevice` for device IDs.

Reply:
Our tile indexing is (I think) all in `int64_t`, so I was following that.  (And my original draft completely removed the `gridinfo` function.)  I can switch it to `int` if that'd be preferable.

Reply:
A related question: I noticed that you use pointers for out parameters while this uses a reference.  Should I change this to use pointer arguments?

Reply:
Tile indices are `int64_t`, MPI and process grid indices are generally `int`. It would avoid the extra variables and cast.

We aren't consistent about use of references. Google's style guide suggests using only `const` references; if the value is an out parameter, use a pointer for clarity. I think that's probably a good rule-of-thumb. But we violate that for `posv_mixed( ..., int& iter, ... )` and of course _all_ matrices.